tensorflow/lite/tools/benchmark/external_delegate_provider.cc

```diff
@@ -0,0 +1,182 @@
+/* Copyright 2020 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+#include "tensorflow/lite/tools/benchmark/benchmark_model.h"
+#include "tensorflow/lite/tools/benchmark/delegate_provider.h"
+#include "tensorflow/lite/tools/benchmark/logging.h"
+
+#if defined(_WIN32)
+#include <Windows.h>
+#else
+#include <dlfcn.h>
+#endif
+
+#include <string>
+#include <type_traits>
+#include <vector>
+
+namespace tflite {
+namespace benchmark {
+namespace {
+// Library Support construct to handle dynamic library operations
+#if defined(_WIN32)
+struct LibSupport {
+  static void* Load(const char* lib) { return LoadLibrary(lib); }
+
+  static void* GetSymbol(void* handle, const char* symbol) {
+    return (void*)GetProcAddress((HMODULE)handle, symbol);
+  }
+
+  static int UnLoad(void* handle) { return FreeLibrary((HMODULE)handle); }
+};
+#else
+struct LibSupport {
+  static void* Load(const char* lib) {
+    return dlopen(lib, RTLD_LAZY | RTLD_LOCAL);
+  }
+
+  static void* GetSymbol(void* handle, const char* symbol) {
+    return dlsym(handle, symbol);
+  }
+
+  static int UnLoad(void* handle) { return dlclose(handle); }
+};
+#endif
+
+// Split a given string to a vector of string using a delimiter character
+std::vector<std::string> SplitString(const std::string& str, char delimiter) {
+  std::vector<std::string> tokens;
+  std::string token;
+  std::istringstream ss(str);
+  while (std::getline(ss, token, delimiter)) {
+    tokens.push_back(token);
+  }
+  return tokens;
+}
+
+// External delegate library construct
+struct ExternalLib {
+  using CreateDelegatePtr = std::add_pointer<TfLiteDelegate*(
+      const char**, const char**, size_t,
+      void (*report_error)(const char*))>::type;
+  using DestroyDelegatePtr = std::add_pointer<void(TfLiteDelegate*)>::type;
+
+  // Open a given delegate library and load the create/destroy symbols
+  void load(const std::string library) {
+    if (!is_loaded) {
```

Comment:
Shouldn't we reload the library here? Seems like the provider could be invoked with different options?

I'd be OK simply loading ExternalLib on the stack in CreateTfLiteDelegate. Looks like we don't use Unload anywhere, and it's fine to leak the handle for benchmarking purposes.

Reply:
Yes I thought the same, leaking the handle should be fine for benchmarking purposes.